### ISSUE TITLE
PHP 8.1 support

### ISSUE THREAD

**Issue by kukulich** [squizlabs/PHP_CodeSniffer#3479]
Issue to track what needs to be done:

- [x] Support of `enum` keyword https://github.com/squizlabs/PHP_CodeSniffer/pull/3478
- [x] `enum` support to sniffs https://github.com/squizlabs/PHP_CodeSniffer/pull/3482
- [x] New token `T_ENUM_CASE` https://github.com/squizlabs/PHP_CodeSniffer/pull/3483
- [x] Support of `readonly` keyword https://github.com/squizlabs/PHP_CodeSniffer/pull/3480
- [x] `never` buildin type https://github.com/squizlabs/PHP_CodeSniffer/pull/3582
- [x] Intersection types - new token `T_TYPE_INTERSECTION` and  update `File::getMethodParameters()`, `File::getMethodProperties()` and `File::getMemberProperties()` https://github.com/squizlabs/PHP_CodeSniffer/pull/3581
- [x] Explicit octal notation https://github.com/squizlabs/PHP_CodeSniffer/pull/3481
- [x] Final constants - reported in PSR12.Properties.ConstantVisibility false positive when using public final const syntax #3526, fixed in PHP 8.1 | PSR12/ConstantVisibility: allow for class constants to be final #3527. A cursory review of existing sniffs showed no other obvious sniffs which would need adujsting.

**Comment by jrfnl** [comment]
**Proposal**: 

For **intersection types**, I'd like to propose to re-tokenize to a new PHPCS native `T_TYPE_INTERSECTION` to prevent sniffs looking for `T_BITWISE_AND` from getting confused and starting to try and handle the token.

This is in-line with the changes made for PHP 8.0 union types (see #3032 and #3302).

Note: the `File::getMethodParameters()`, `File::getMethodProperties()` and `File::getMemberProperties()` methods will also need adjusting.

---

Also note that the `File::getMemberProperties()` also still needs adjusting for the `readonly` keyword.

**Comment by jrfnl** [comment]
~~Check sniffs that use Tokens::$ooScopeTokens~~

For all `Tokens` arrays being updated with new tokens:
- [ ] Check every single PHPCS native sniff using them for either handling the new syntax correctly or bowing out.
- [ ] No matter what the decision: add a couple of tests to each of these sniffs to guard against the sniff throwing false positives for the new syntax/fixing things incorrectly etc.

- [ ] Even when not using the `Tokens` arrays, see if sniffs looking for "similar" tokens need to add support for the new syntaxes.
    Think for example: sniffs which looks at type declarations in a function declaration will now also need to handle intersection types and `never`. In most cases, those sniffs will use the `File::getMethodParameters()` and `File::getMethodProperties()`, so instead of doing a token search, do a search for usage of these methods and review each sniff using them.

**Comment by MarkBaker** [comment]
Addition to the list:
 - Support for first class callable syntax

**Comment by jrfnl** [comment]
> At the moment, this tokenises to T_ELLIPSIS; but it might be a good idea to change that to a new token (T_FIRST_CLASS_CALLABLE) which will then make the usage explicit

For the purpose of writing sniffs, I think it is completely unnecessary to retokenize that. Use as a first class callable is straight forward to detect (is the next non-empty token a close parenthesis ?).

Retokenizing to a PHPCS native token constant is generally only done when it would need _a lot_ of logic within a sniff to determine whether the token is the target token, or when not retokenizing would lead to lots of false positives in pre-existing sniffs.

Neither applies.

**Comment by jrfnl** [comment]
To add to the list:

* [Final constants](https://wiki.php.net/rfc/final_class_const) - reported in #3526, fixed in #3527. A cursory review of existing sniffs showed no other obvious sniffs which would need adujsting.

**Comment by nunoperalta** [comment]
I was about to report the same as #557

enum being detected as GOTO:

```php
enum Suit: string {
	case Hearts = 'H';
	case Diamonds = 'D';
	case Clubs = 'C';
	case Spades = 'S';
}
```

> Use of the GOTO language construct is discouragedPHPCS(Generic.PHP.DiscourageGoto.Found)

**Comment by jrfnl** [comment]
@nunoperalta There is already a PR open to fix that: #3534

**Comment by nunoperalta** [comment]
@jrfnl nice one! Any idea when it will be merged? I see it's been there for 15 days. Thank you very much.

**Comment by jrfnl** [comment]
> nice one! Any idea when it will be merged?

@nunoperalta When the maintainers have time to review.

**Comment by acidjazz** [comment]
👀 @jrfnl @gsherwood @michalbundyra @gmponos 👀

**Comment by acidjazz** [comment]
@gsherwood @jrfnl @kukulich loving and using Enums in all of my current projects

is it possible at all to separate the Enum support and get that reviewed/merged?

**Comment by kukulich** [comment]
Everything is solved now in `master`.

**Comment by mreduar** [comment]
Thank you!!
